lib/runner/variant-reassignment.js

```diff
@@ -17,52 +16,148 @@ export default class VariantReassignment {
+   * @returns {Promise.<*>}
+   */
   async execute (productDrafts, existingProductProjections) {
-    this._processUnfinishedTransactions()
+    let products
+
+    try {
+      if (this.firstRun)
+        await this._processUnfinishedTransactions()
+    } catch (e) {
+      return this._error('Could not process unfinished transactions', e)
+    }
+    this.firstRun = false
 
-    const products
-      = await this.productService.fetchProductsFromProductProjections(existingProductProjections)
+    try {
+      products
+        = await this.productService.fetchProductsFromProductProjections(existingProductProjections)
+    } catch (e) {
+      return this._error('Error while fetching products for reassignment', e)
+    }
 
     const productDraftsForReassignment
       = this._selectProductDraftsForReassignment(productDrafts, products)
 
-    if (productDraftsForReassignment.length)
-      for (const productDraft of productDraftsForReassignment)
+    this.logger.debug(
+      'Filtered %d productDrafts for reassignment',
+      productDraftsForReassignment.length
+    )
+
+    for (const productDraft of productDraftsForReassignment)
+      try {
         await this._processProductDraft(productDraft, products)
+      } catch (e) {
+        this.logger.error(
+          'Error while processing productDraft %j, retrying.',
+          productDraft.name, e.stack
+        )
+        await this._handleProcessingError(productDraft, products)
+      } finally {
+        this.logger.debug(
+          'Finished processing of productDraft with name %j',
+          productDraft.name
+        )
+      }
+
+    return Promise.resolve()
   }
 
-  async _processUnfinishedTransactions () {
-    if (this.firstRun)
-      this.unfinishedTransactions = [] // API.getUnfinishedTransactions()
+  async _handleProcessingError (productDraft, products) {
+    const transactions = await this.transactionService.getTransactions()
+    const failedTransaction = transactions.find(({ value }) =>
+      _.isEqual(value.newProductDraft.name, productDraft.name)
+    )
 
-    this.firstRun = false
-    for (const transaction of this.unfinishedTransactions)
-      await this._createAndExecuteActions(transaction.newProductDraft,
-        transaction.backupProductDraft, transaction.variants, transaction.ctpProductToUpdate)
+    return failedTransaction
+      // transaction was created but not finished, try to finish it
+      ? this._processUnfinishedTransactions(transactions)
+      // transaction was not created, try to process productDraft again
+      : this._processProductDraft(productDraft, products)
+  }
+
+  /**
+   * Log error and return Promise.reject
+   * @param msg String with error description
+   * @param e Error object with details
+   * @return <Promise.reject>
+   * @private
+   */
+  _error (msg, e) {
+    const error = (e && e.message) || String(e)
+    this.logger.error(msg, e)
+    return Promise.reject(new Error(`${msg} - ${error}`))
```

Comment:
This was again because of a linter error [prefer-promise-reject-errors](https://eslint.org/docs/rules/prefer-promise-reject-errors). But it does not do it again so maybe it was disabled somewhere? Should I change it? wdyt?

Reply:
It's ok, let's leave it like this for now. Let's see if it will make any problems during integration.